### PR TITLE
add support for meta transactions

### LINKED ISSUE #98: 3rd party funding
The docs for `ethr-did-registry` mention this:

> Since each Ethereum transaction has to be funded, there is a growing trend of on-chain transactions that are authenticated via an externally created signature and not by the actual transaction originator. This allows 3rd party funding services or receiver pays without any fundamental changes to the underlying Ethereum architecture.

I assume this is done by using the `*Signed` version of the methods like `setAttributeSigned`, but I couldn't find support for this in this library nor in the `ethr-did-resolver`.

- Is there any example about this?
- What would it take to implement this feature?

Thanks!

### PR DESCRIPTION
**Pull request by strumswell**
fixes #98

With ethr-did-resolver getting its 6.2.0 update, I updated this dependency to reflect the underlying additions. This includes wrapping the create hash function and the corresponding signed methods doing the txs. Also added tests for it.

Let me know what you think!

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -198,6 +296,33 @@ export class EthrDID {
...
+    }
+    return this.controller.createRevokeAttributeHash(attrName, attrValue)
+  }
+
+  async revokeAttributeSigned(
+    key: string,
+    value: string | Uint8Array,
+    signature: MetaSignature,
+    /** @deprecated please use `txOptions.gasLimit` */
+    gasLimit?: number,
```
Comment (mirceanis) [inline src/index.ts:311]: We shouldn't be adding deprecated params to new methods.

File: package.json [hunk 2]
```diff
@@ -85,6 +85,6 @@
     "@ethersproject/wallet": "^5.6.2",
     "did-jwt": "^6.3.0",
     "did-resolver": "^4.0.0",
-    "ethr-did-resolver": "^6.1.0"
+    "ethr-did-resolver": "^6.2.0"
```
Comment (mirceanis) [inline package.json:88]: This should be bumped to ^6.2.1 now

**Review by mirceanis — CHANGES_REQUESTED** [review]
Thanks for contributing this!
Let's bump the resolver and remove deprecated params before merging.

File: src/index.ts [hunk 3]
```diff
@@ -198,6 +296,33 @@ export class EthrDID {
...
+      throw new Error('a web3 provider configuration is needed for network operations')
+    }
+    return this.controller.createRevokeAttributeHash(attrName, attrValue)
+  }
+
+  async revokeAttributeSigned(
+    key: string,
+    value: string | Uint8Array,
+    signature: MetaSignature,
+    /** @deprecated please use `txOptions.gasLimit` */
```
[the remark this answers is not in the document]
Comment (strumswell) [reply src/index.ts]: Removed it.

File: package.json (hunk 2 above)
Comment (strumswell) [reply package.json]: Done 👍🏻

**Comment by strumswell** [comment]
Made the requested changes! Are we good now?

**Comment by mirceanis** [comment]
> Made the requested changes! Are we good now?

Almost :)

The `gasLimit` parameter that is deprecated in existing methods should not be copied to the new `*Signed` methods.
Users should be using `txOptions.gasLimit` instead.
Please add the `@deprecated` hints back for the existing methods and remove the `gasLimit` parameter from the new methods.

File: src/index.ts [hunk 4]
```diff
@@ -11,6 +11,7 @@ import { Base58 } from '@ethersproject/basex'
 import { toUtf8Bytes } from '@ethersproject/strings'
 import { EthrDidController, interpretIdentifier, REGISTRY } from 'ethr-did-resolver'
 import { Resolvable } from 'did-resolver'
+import { address, MetaSignature } from 'ethr-did-resolver/lib/helpers'
```
Comment (strumswell) [inline src/index.ts:14]: I have also noticed that we have not exported these in `ethr-did-resolver` and that it is breaking your pipeline. I have not much experience with writing packages so what do we do about it? I guess we can either redeclare them here or export them in the `index.ts` (?) in `ethr-did-resolver`. Wdyt?

**Comment by strumswell** [comment]
> > Made the requested changes! Are we good now?
> 
> Almost :)
> 
> The `gasLimit` parameter that is deprecated in existing methods should not be copied to the new `*Signed` methods. Users should be using `txOptions.gasLimit` instead. Please add the `@deprecated` hints back for the existing methods and remove the `gasLimit` parameter from the new methods.

My bad, thanks for pointing this out. 😅

File: src/index.ts (hunk 4 above)
Comment (mirceanis) [inline src/index.ts:14]: Looks like this import is failing. These types are simple enough to be redefined here until they get exported from the resolver package:

```suggestion

export type address = string
export type MetaSignature = {
  sigV: number
  sigR: bytes32
  sigS: bytes32
}
```

File: src/index.ts [hunk 5]
```diff
@@ -179,6 +249,33 @@ export class EthrDID {
...
+    }
+    return this.controller.createSetAttributeHash(attrName, attrValue, exp)
+  }
+
+  async setAttributeSigned(
+    key: string,
+    value: string | Uint8Array,
+    expiresIn = 86400,
+    signature: MetaSignature,
+    gasLimit?: number,
```
Comment (mirceanis) [inline src/index.ts:264]: Let's remove the `gasLimit` parameter. It should be extracted from the `txOptions`
```suggestion
```

File: src/index.ts [hunk 6]
```diff
@@ -146,6 +165,33 @@ export class EthrDID {
...
+      throw new Error('a web3 provider configuration is needed for network operations')
+    }
+    const owner = await this.lookupOwner()
+    const receipt = await this.controller.addDelegateSigned(
+      delegateOptions?.delegateType || DelegateTypes.veriKey,
+      delegate,
+      delegateOptions?.expiresIn || 86400,
+      signature,
+      { ...txOptions, from: owner }
+    )
```
Comment (mirceanis) [inline src/index.ts:191]: The `*Signed()` methods don't require the sender to be the current owner.

```suggestion
    const receipt = await this.controller.addDelegateSigned(
      delegateOptions?.delegateType || DelegateTypes.veriKey,
      delegate,
      delegateOptions?.expiresIn || 86400,
      signature,
      txOptions
    )
```

File: src/index.ts [hunk 7]
```diff
@@ -159,11 +205,35 @@ export class EthrDID {
...
+    txOptions: CallOverrides = {}
+  ): Promise<string> {
+    if (typeof this.controller === 'undefined') {
+      throw new Error('a web3 provider configuration is needed for network operations')
+    }
+    const owner = await this.lookupOwner()
+    const receipt = await this.controller.revokeDelegateSigned(delegateType, delegate, signature, {
+      ...txOptions,
+      from: owner,
+    })
```
Comment (mirceanis) [inline src/index.ts:228]:
```suggestion
    const receipt = await this.controller.revokeDelegateSigned(delegateType, delegate, signature, txOptions)
```

File: src/index.ts [hunk 8]
```diff
@@ -179,6 +249,33 @@ export class EthrDID {
...
+  ): Promise<string> {
+    if (typeof this.controller === 'undefined') {
+      throw new Error('a web3 provider configuration is needed for network operations')
+    }
+    const owner = await this.lookupOwner()
+    const receipt = await this.controller.setAttributeSigned(key, attributeToHex(key, value), expiresIn, signature, {
+      gasLimit,
+      ...txOptions,
+      from: owner,
+    })
```
Comment (mirceanis) [inline src/index.ts:275]:
```suggestion
    const receipt = await this.controller.setAttributeSigned(key, attributeToHex(key, value), expiresIn, signature, txOptions)
```

File: src/index.ts [hunk 9]
```diff
@@ -198,6 +295,32 @@ export class EthrDID {
...
+      throw new Error('a web3 provider configuration is needed for network operations')
+    }
+    const owner = await this.lookupOwner()
+    const receipt = await this.controller.revokeAttributeSigned(key, attributeToHex(key, value), signature, {
+      gasLimit,
+      ...txOptions,
+      from: owner,
+    })
+    return receipt.transactionHash
+  }
```
Comment (mirceanis) [inline src/index.ts:322]:
```suggestion
  async revokeAttributeSigned(
    key: string,
    value: string | Uint8Array,
    signature: MetaSignature,
    txOptions: CallOverrides = {}
  ): Promise<string> {
    if (typeof this.controller === 'undefined') {
      throw new Error('a web3 provider configuration is needed for network operations')
    }
    const receipt = await this.controller.revokeAttributeSigned(key, attributeToHex(key, value), signature, txOptions)
    return receipt.transactionHash
  }
```

File: src/index.ts [hunk 10]
```diff
@@ -128,6 +129,24 @@ export class EthrDID {
...
+    return this.controller.createChangeOwnerHash(newOwner)
+  }
+
+  async changeOwnerSigned(newOwner: string, signature: MetaSignature, txOptions?: CallOverrides): Promise<string> {
+    if (typeof this.controller === 'undefined') {
+      throw new Error('a web3 provider configuration is needed for network operations')
+    }
+    const receipt = await this.controller.changeOwnerSigned(newOwner, signature, {
+      ...txOptions,
+    })
```
Comment (mirceanis) [inline src/index.ts:145]:
```suggestion
    const receipt = await this.controller.changeOwnerSigned(newOwner, signature, txOptions)
```

**Review by mirceanis — CHANGES_REQUESTED** [review]
There's an import that's failing and I also made some suggestions about the `gasLimit` parameter and the transaction sender in `txOptions`

File: src/index.ts (hunk 4 above)
Comment (mirceanis) [reply src/index.ts:14]: It's ok to redeclare until they get exported from there.. they are simple enough types, and unlikely to change and cause a conflict

**Comment by strumswell** [comment]
Thank you so much for taking your time going through this! I had some time now and included all your suggestions but:
- Didn't including the `MetaTransaction` type. I saw you already released a new version of the resolver that exports it. So bumped its version in the package and imported from there.
- Made `txOptions` parameter in `changeOwnerSigned` non-optional to be in line with the other signed methods.
- Removed usage of `address`. You are not using this at all in other places so I just resolved them back to their underlying `string` type. Then we also don't have to export this type anywhere.

**Review by mirceanis — APPROVED** [review]
Looks great, thanks for contributing this!

**Pull request merged**